### PR TITLE
Intly 2798 delete users and federated identity

### PR DESCRIPTION
**Pull request by sedroche**
## Motivation
Jira: https://issues.jboss.org/browse/INTLY-2798

## What
As part of the above ticket this PR adds functionality to:
- ~`deleteUsers` option as part of the KeycloakRealm. Currently deleting users on the CR is not mapped into the keycloak users. Default is `false` for this property~
After convo in the comments deleting users is reconciled based on `createOnly: false`
- `federatedIdentities` property as part of the User spec. This is used to map federated identities from identity providers.

## Why
Jira: https://issues.jboss.org/browse/INTLY-2798

## How
Extended the `reconcileUser` function to handle the above features.
Also adding:
```go
CreateFederatedIdentity(fid v1alpha1.FederatedIdentity, userId string, realmName string) error
RemoveFederatedIdentity(fid v1alpha1.FederatedIdentity, userId string, realmName string) error
GetUserFederatedIdentities(userName string, realmName string) ([]v1alpha1.FederatedIdentity, error)
```
to the client
## Verification Steps
Can be verified as part of https://github.com/integr8ly/integreatly-operator/pull/81

## Checklist:

- [x] Code has been tested locally by PR requester
- [ ] Changes have been successfully verified by another team member

### CODE REVIEW

File: pkg/keycloak/realm/phaseHandler.go [hunk 1]
```diff
@@ -281,7 +289,46 @@ func (ph *phaseHandler) reconcileUserClientRoles(specUser *v1alpha1.KeycloakUser
...
+	}
+	for _, specFid := range specUser.FederatedIdentities {
+		if !containsFi(currentFids, specFid) {
+			me.AddError(authenticatedClient.CreateFederatedIdentity(specFid, user.ID, realmName))
+		}
+	}
+	if me.IsNil() {
+		return nil
+	}
+	return nil
```
Comment (matskiv) [inline pkg/keycloak/realm/phaseHandler.go]: Shouldn't we return error at this point?

**Comment by matskiv** [comment]
I think we should make user deletion implicit behavior when `createOnly` is false, and avoid another flag - `deleteUsers`.
@sedroche @philbrookes wdyt?

File: pkg/keycloak/realm/phaseHandler.go (hunk 1 above)
Comment (sedroche) [reply pkg/keycloak/realm/phaseHandler.go]: Good spot. Ill update this now :+1:

**Comment by philbrookes** [comment]
I believe the situation here is:
`createOnly: true` create any entities are missing and nothing else
`createOnly: false` all changes to the CR are reconciled each reconcile loop, users not in CR are not deleted.
`createOnly: false && deleteUsers: true` all changes to the CR are reconciled each reconcile loop, users not in CR are also deleted.

I think 2 flags are required to stop `createOnly: false` from incorrectly deleting users created through user registration against keycloak.

**Comment by matskiv** [comment]
@philbrookes AFAIK in Integreatly we don't have a use case where we would have `createOnly: false && deleteUsers: false`. That's why I don't see a benefit in having `deleteUsers` flag.

**Comment by philbrookes** [comment]
correct: 
`createOnly: false && deleteUsers: false` is the same as `createOnly: false`
`createOnly: true` is the same as `createOnly: true && deleteUsers: false`
however, with only `createOnly`, how do we get: `createOnly: false && deleteUsers: true`?

**Comment by philbrookes** [comment]
@sedroche isn't the `createOnly: false && deleteUsers: false` the situation you need?

**Comment by matskiv** [comment]
> however, with only createOnly, how do we get: createOnly: false && deleteUsers: true?

@philbrookes I am suggesting to make user deletion implicit behavior of `createOnly: false`. Basically if `createOnly: false` - fully manage realm by the CR.
But one more flag is not a big deal. It's just a suggestion.

**Comment by sedroche** [comment]
@philbrookes I am using `createOnly: false`  `deleteUsers: true`
The reasoning for this was that although `createOnly` is supposed to reconcile all changes to the CR, Users were not deleted because that would delete users created by an admin through the UI and we didn't want that behaviour.
`deleteUsers` says to specifically do this.

**Comment by philbrookes** [comment]
if we have no use-case where we want the CR reconciled by the operator in all cases except deleting the users, then let's remove that flag.

**Pull request merged**